File: onedal/decomposition/pca.cpp
/*******************************************************************************
* Copyright 2023 Intel Corporation
*
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
*     http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*******************************************************************************/
#include "oneapi/dal/algo/pca.hpp"
#include "onedal/common.hpp"

namespace py = pybind11;

namespace oneapi::dal::python {
namespace decomposition {
struct params2desc {
    template <typename Float, typename Method, typename Task>
    auto operator()(const pybind11::dict& params) {
        const auto n_components = params["n_components"].cast<std::int64_t>();
        bool whiten = params["whiten"].cast<bool>();
        // sign-flip feature is always used in scikit-learn
        bool is_deterministic = params["is_deterministic"].cast<bool>();

        auto desc = pca::descriptor<Float, Method>()
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240100
                        .set_whiten(whiten)
                        .set_normalization_mode(dal::pca::normalization::mean_center)
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION>=20240100
                        .set_component_count(n_components)
                        .set_deterministic(is_deterministic);
        return desc;
    }
};

template <typename Task, typename Ops>
struct method2t {
    method2t(const Task& task, const Ops& ops) : ops(ops) {}

    template <typename Float>
    auto operator()(const py::dict& params) {
        using namespace dal::pca;

        const auto method = params["method"].cast<std::string>();
        ONEDAL_PARAM_DISPATCH_VALUE(method, "cov", ops, Float, method::cov);
        ONEDAL_PARAM_DISPATCH_VALUE(method, "svd", ops, Float, method::svd);
        ONEDAL_PARAM_DISPATCH_VALUE(method, "precomputed", ops, Float, method::precomputed);
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
    }

    Ops ops;
};

template <typename Task, typename Ops>
struct incrementalmethod2t {
    incrementalmethod2t(const Task& task, const Ops& ops) : ops(ops) {}

    template <typename Float>
    auto operator()(const py::dict& params) {
        using namespace dal::pca;

        const auto method = params["method"].cast<std::string>();
        ONEDAL_PARAM_DISPATCH_VALUE(method, "cov", ops, Float, method::cov);
        ONEDAL_PARAM_DISPATCH_THROW_INVALID_VALUE(method);
    }

    Ops ops;
};

template <typename Task>
void init_model(py::module_& m) {
    using namespace dal::pca;
    using model_t = model<Task>;

    auto cls = py::class_<model_t>(m, "model")
                   .def(py::init())
                   .def(py::pickle(
                       [](const model_t& m) {
                           return serialize(m);
                       },
                       [](const py::bytes& bytes) {
                           return deserialize<model_t>(bytes);
                       }))
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240100
                   .DEF_ONEDAL_PY_PROPERTY(eigenvalues, model_t)
                   .DEF_ONEDAL_PY_PROPERTY(means, model_t)
                   .DEF_ONEDAL_PY_PROPERTY(variances, model_t)
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION>=20240100
                   .DEF_ONEDAL_PY_PROPERTY(eigenvectors, model_t);
}

template <typename Task>
void init_train_result(py::module_& m) {
    using namespace dal::pca;
    using result_t = train_result<Task>;

    py::class_<result_t>(m, "train_result")
        .def(py::init())
        .DEF_ONEDAL_PY_PROPERTY(model, result_t)
        .def_property_readonly("eigenvectors", &result_t::get_eigenvectors)
        .def_property_readonly("eigenvalues", &result_t::get_eigenvalues)
#if defined(ONEDAL_VERSION) && ONEDAL_VERSION >= 20240100
        .def_property_readonly("singular_values", &result_t::get_singular_values)
        .def_property_readonly("explained_variances_ratio", &result_t::get_explained_variances_ratio)
#endif // defined(ONEDAL_VERSION) && ONEDAL_VERSION>=20240100
        .def_property_readonly("means", &result_t::get_means)
        .def_property_readonly("variances", &result_t::get_variances);
}

template <typename Task>
void init_partial_train_result(py::module_& m) {
    using namespace dal::pca;
    using result_t = partial_train_result<Task>;

    py::class_<result_t>(m, "partial_train_result")
        .def(py::init())
        .DEF_ONEDAL_PY_PROPERTY(partial_n_rows, result_t)
        .DEF_ONEDAL_PY_PROPERTY(partial_crossproduct, result_t)
        .DEF_ONEDAL_PY_PROPERTY(partial_sum, result_t)
        .DEF_ONEDAL_PY_PROPERTY(auxiliary_table, result_t);
}

template <typename Task>
void init_infer_result(py::module_& m) {
    using namespace dal::pca;
    using result_t = infer_result<Task>;

    auto cls = py::class_<result_t>(m, "infer_result")
                   .def(py::init())
                   .DEF_ONEDAL_PY_PROPERTY(transformed_data, result_t);
}

template <typename Policy, typename Task>
void init_train_ops(py::module& m) {
    m.def("train", [](const Policy& policy, const py::dict& params, const table& data) {
        using namespace dal::pca;
        using input_t = train_input<Task>;

        train_ops ops(policy, input_t{ data }, params2desc{});
        return fptype2t{ method2t{ Task{}, ops } }(params);
    });
}

template <typename Policy, typename Task>
void init_partial_train_ops(py::module& m) {
    using prev_result_t = dal::pca::partial_train_result<Task>;
    m.def("partial_train", [](
        const Policy& policy,
        const py::dict& params,
        const prev_result_t& prev,
        const table& data) {
            using namespace dal::pca;
            using input_t = partial_train_input<Task>;
            partial_train_ops ops(policy, input_t{ prev, data }, params2desc{});
            return fptype2t{ incrementalmethod2t{ Task{}, ops } }(params);
        }
    );
};

template <typename Policy, typename Task>
void init_finalize_train_ops(py::module& m) {
    using input_t = dal::pca::partial_train_result<Task>;
    m.def("finalize_train", [](const Policy& policy, const py::dict& params, const input_t& data) {
            finalize_train_ops ops(policy, data, params2desc{});
            return fptype2t{ incrementalmethod2t{ Task{}, ops } }(params);
        }
    );
};

template <typename Policy, typename Task>
void init_infer_ops(py::module_& m) {
    m.def("infer",
          [](const Policy& policy,
             const py::dict& params,
             const pca::model<Task>& model,
             const table& data) {
              using namespace dal::pca;
              using input_t = infer_input<Task>;

              infer_ops ops(policy, input_t{ model, data }, params2desc{});
              return fptype2t{ method2t{ Task{}, ops } }(params);
          });
}

ONEDAL_PY_DECLARE_INSTANTIATOR(init_model);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_partial_train_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_result);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_partial_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_finalize_train_ops);
ONEDAL_PY_DECLARE_INSTANTIATOR(init_infer_ops);
} // namespace decomposition

ONEDAL_PY_INIT_MODULE(decomposition) {
    using namespace decomposition;
    using namespace dal::pca;
    using namespace dal::detail;

    using task_list = types<task::dim_reduction>;
    auto sub = m.def_submodule("decomposition");
    #ifdef ONEDAL_DATA_PARALLEL_SPMD
        ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_spmd, task_list);
    #else  
        ONEDAL_PY_INSTANTIATE(init_train_ops, sub, policy_list, task_list);
        ONEDAL_PY_INSTANTIATE(init_infer_ops, sub, policy_list, task_list);
        ONEDAL_PY_INSTANTIATE(init_model, sub, task_list);
        ONEDAL_PY_INSTANTIATE(init_train_result, sub, task_list);
        ONEDAL_PY_INSTANTIATE(init_partial_train_result, sub, task_list);
        ONEDAL_PY_INSTANTIATE(init_infer_result, sub, task_list);
        ONEDAL_PY_INSTANTIATE(init_partial_train_ops, sub, policy_list, task_list);
        ONEDAL_PY_INSTANTIATE(init_finalize_train_ops, sub, policy_list, task_list);
    #endif
}

ONEDAL_PY_TYPE2STR(dal::pca::task::dim_reduction, "dim_reduction");
} //namespace oneapi::dal::python

File: onedal/decomposition/pca.py
# ==============================================================================
# Copyright 2023 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

import numbers
from abc import ABCMeta

import numpy as np
from sklearn.decomposition._pca import _infer_dimension
from sklearn.utils.extmath import stable_cumsum

from ..common._base import BaseEstimator
from ..datatypes import _convert_to_supported, from_table, to_table
from ..utils._array_api import (
    _asarray,
    _convert_to_numpy,
    get_namespace,
    sklearn_array_api_dispatch,
)


class BasePCA(BaseEstimator, metaclass=ABCMeta):
    """
    Base class for PCA oneDAL implementation.
    """

    def __init__(
        self,
        n_components=None,
        is_deterministic=True,
        method="cov",
        whiten=False,
    ):
        self.n_components = n_components
        self.method = method
        self.is_deterministic = is_deterministic
        self.whiten = whiten

    def _get_onedal_params(self, data, xp, stage=None):
        if stage is None:
            n_components = self._resolve_n_components_for_training(data.shape)
        elif stage == "predict":
            n_components = self.n_components_
        return {
            "fptype": "float" if data.dtype == xp.float32 else "double",
            "method": self.method,
            "n_components": n_components,
            "is_deterministic": self.is_deterministic,
            "whiten": self.whiten,
        }

    def _validate_n_components(self, n_components, n_samples, n_features):
        if n_components is None:
            n_components = min(n_samples, n_features)
        if n_components == "mle":
            if n_samples < n_features:
                raise ValueError(
                    "n_components='mle' is only supported if n_samples >= n_features"
                )
        elif not 0 <= n_components <= min(n_samples, n_features):
            raise ValueError(
                "n_components=%r must be between 0 and "
                "min(n_samples, n_features)=%r with "
                "svd_solver='full'" % (n_components, min(n_samples, n_features))
            )
        elif n_components >= 1:
            if not isinstance(n_components, numbers.Integral):
                raise ValueError(
                    "n_components=%r must be of type int "
                    "when greater than or equal to 1, "
                    "was of type=%r" % (n_components, type(n_components))
                )

    def _resolve_n_components_for_training(self, shape_tuple):
        if self.n_components is None or self.n_components == "mle":
            return min(shape_tuple)
        elif (
            isinstance(self.n_components, float)
            and self.n_components > 0.0
            and self.n_components <= 1.0
        ):
            return min(shape_tuple)
        else:
            return self.n_components

    def _resolve_n_components_for_result(self, shape_tuple):
        if self.n_components is None:
            return min(shape_tuple)
        elif self.n_components == "mle":
            return _infer_dimension(self.explained_variance_, shape_tuple[0])
        elif 0.0 < self.n_components < 1.0:
            # TODO:
            # check for Array API.
            ratio_cumsum = stable_cumsum(self.explained_variance_ratio_)
            return np.searchsorted(ratio_cumsum, self.n_components, side="right") + 1
        elif isinstance(self.n_components, float) and self.n_components == 1.0:
            return min(shape_tuple)
        else:
            return self.n_components

    def _compute_noise_variance(self, xp, n_components, n_sf_min):
        if n_components < n_sf_min:
            if len(self.explained_variance_) == n_sf_min:
                return self.explained_variance_[n_components:].mean()
            elif len(self.explained_variance_) < n_sf_min:
                # TODO Rename variances_ to var_ to align with sklearn/sklearnex IncrementalPCA
                # TODO:
                # check xp.sum for Array API.
                if hasattr(self, "variances_"):
                    resid_var = xp.sum(self.variances_)
                elif hasattr(self, "var_"):
                    resid_var = xp.sum(self.var_)

                resid_var -= xp.sum(self.explained_variance_)
                return resid_var / (n_sf_min - n_components)
        else:
            return 0.0

    def _create_model(self, xp):
        m = self._get_backend("decomposition", "dim_reduction", "model")
        m.eigenvectors = to_table(_convert_to_numpy(self.components_, xp=xp))
        m.means = to_table(_convert_to_numpy(self.mean_, xp=xp))
        if self.whiten:
            m.eigenvalues = to_table(_convert_to_numpy(self.explained_variance_, xp=xp))
        self._onedal_model = m
        return m

    def _predict(self, X, xp, queue=None):
        policy = self._get_policy(queue, X)
        model = self._create_model(xp)
        X = _convert_to_supported(policy, X)
        params = self._get_onedal_params(X, xp, stage="predict")

        result = self._get_backend(
            "decomposition",
            "dim_reduction",
            "infer",
            policy,
            params,
            model,
            to_table(_convert_to_numpy(X, xp=xp)),
        )
        # Since `from_table` data management enabled only for numpy host,
        # copy data from numpy host output to xp namespace array.
        return _asarray(
            from_table(result.transformed_data).reshape(-1), xp=xp, sycl_queue=queue
        )

    def _predict(self, X, xp, queue=None):
        xp, is_array_api_compliant = get_namespace(X)
        # update for queue getting.
        queue = X.sycl_queue
        return self._fit(X, xp, is_array_api_compliant, queue)


class PCA(BasePCA):

    @sklearn_array_api_dispatch()
    def _fit(self, X, xp, is_array_api_compliant, y=None, queue=None):
        n_samples, n_features = X.shape
        n_sf_min = min(n_samples, n_features)
        self._validate_n_components(self.n_components, n_samples, n_features)

        policy = self._get_policy(queue, X)
        # TODO: investigate why np.ndarray with OWNDATA=FALSE flag
        # fails to be converted to oneDAL table
        # TODO:
        # check if only numpy issues.
        if isinstance(X, np.ndarray) and not X.flags["OWNDATA"]:
            X = X.copy()
        X = _convert_to_supported(policy, X, xp=xp)

        params = self._get_onedal_params(X)
        result = self._get_backend(
            "decomposition",
            "dim_reduction",
            "train",
            policy,
            params,
            to_table(_convert_to_numpy(X, xp=xp)),
        )

        # Since `from_table` data management enabled only for numpy host,
        # copy data from numpy host output to xp namespace array.
        self.mean_ = _asarray(
            from_table(result.means).reshape(-1), xp=xp, sycl_queue=queue
        )
        self.variances_ = _asarray(
            from_table(result.variances).reshape(-1), xp=xp, sycl_queue=queue
        )
        self.components_ = _asarray(
            from_table(result.eigenvectors).reshape(-1), xp=xp, sycl_queue=queue
        )
        self.singular_values_ = _asarray(
            from_table(result.singular_values).reshape(-1), xp=xp, sycl_queue=queue
        )
        # self.explained_variance_ = np.maximum(from_table(result.eigenvalues).ravel(), 0)
        # TODO:
        # check for Array API.
        self.explained_variance_ = xp.max(
            _asarray(
                from_table(result.singular_values).reshape(-1), xp=xp, sycl_queue=queue
            ),
            0,
        )
        self.explained_variance_ratio_ = _asarray(
            from_table(result.explained_variances_ratio).reshape(-1),
            xp=xp,
            sycl_queue=queue,
        )
        self.n_samples_ = n_samples
        self.n_features_ = n_features

        U = None
        S = self.singular_values_
        Vt = self.components_

        n_components = self._resolve_n_components_for_result(X.shape)
        self.n_components_ = n_components
        self.noise_variance_ = self._compute_noise_variance(xp, n_components, n_sf_min)

        # TODO:
        # check ufunc work here.
        if n_components < params["n_components"]:
            self.explained_variance_ = self.explained_variance_[:n_components]
            self.components_ = self.components_[:n_components]
            self.singular_values_ = self.singular_values_[:n_components]
            self.explained_variance_ratio_ = self.explained_variance_ratio_[:n_components]

        return self

    def fit(self, X, y=None, queue=None):
        xp, is_array_api_compliant = get_namespace(X)
        # update for queue getting.
        queue = X.sycl_queue
        return self._fit(X, xp, is_array_api_compliant, y, queue)
